R/teal_slice.R

```diff
@@ -268,22 +280,68 @@ c.teal_slice <- function(...) {
   ans
 }
 
+#' @param x `teal_slice` object
+#' @keywords internal
+to_list <- function(x){
+  checkmate::assert_class(x, "teal_slice")
+
+  x <- if (shiny::isRunning()) {
+    shiny::reactiveValuesToList(x)
+  } else {
+    shiny::isolate(shiny::reactiveValuesToList(x))
+  }
+
+  formals_names <- setdiff(names(formals(filter_var)), '...')
+
+  extra_names   <- setdiff(names(x), formals_names)
+  missing_names <- setdiff(formals_names, names(x))
+
+
+  x <- c(x, formals(filter_var)[missing_names])
+
+  x[c(formals_names, extra_names)]
+
+}
+
+#' @param x `list` object
+#' @param ... further arguments passed to [jsonlite::toJSON]
+#' @keywords internal
+to_json <- function(x, ...) {
+  checkmate::assert_class(x, "list")
+  jsonlite::toJSON(x, pretty = TRUE, auto_unbox = TRUE, digits = 16, ...)
+}
 
 # format method for teal_slice
 #' @export
+#' @param x `teal_slice` object
+#' @param show_all `logical(1)` should parameters set to NULL be returned
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slice <- function(x, show_all = FALSE, ...) {
+format.teal_slice <- function(x, show_all = FALSE, pretty = TRUE, ...) {
   checkmate::assert_flag(show_all)
+  checkmate::assert_flag(pretty)
 
-  x <- if (shiny::isRunning()) {
-    rev(shiny::reactiveValuesToList(x))
-  } else {
-    rev(shiny::isolate(shiny::reactiveValuesToList(x)))
-  }
+  x_json <- to_json(to_list(x))
+  x_json_c <- capture.output(print(x_json))
+
+  if (!show_all) x_json_c <- grep('{}', x_json_c, fixed = TRUE, invert = TRUE, value = TRUE)
```

Comment:
Is it not better to run `Filter(Negate(is.null), x)` before converting to JSON?

Reply:
@chlebowa I wrote a long comment on why NOT, and then realized this would look cleaner if it's moved to `to_list` where I specially restore NULL values that are omitted by `filter_var`. So with `show_all = FALSE` i will not restore those values, and with `show_all = TRUE` I will pass it to `to_list` that will restore NULLs

Reply:
fixed that in ad457d7aec5e04e81c3fa1809faf70df130a971c